js/ui/popup.js

```diff
@@ -60,7 +60,14 @@ Popup.prototype = util.inherit(Evented, /** @lends Popup.prototype */{
         this._map = map;
         this._map.on('move', this._update);
         if (this.options.closeOnClick) {
-            this._map.on('click', this._onClickClose);
+            // We attach the `Map#click` listener asynchronously so that it is
+            // possible to open a popup within a `click` event on a child of
+            // the map's container. If we attached the listener syncronously
+            // the popup would close when the click event propogated to `map`.
+            var that = this;
+            setTimeout(function() {
```

Comment:
Good solution! It will probably be a performance improvement for lots of markers because adding native event listeners tends to be slower than event delegation.